candidate_scraper.py

```diff
@@ -0,0 +1,398 @@
+                  }
+        
+    max_number_searched = 100  #max number of pages searched for each domain. 
+        
+    def __init__(self,url,keyword_dict):
+        url = url.strip('/')
+        self.init_url = url
+        self.keyword_dict = keyword_dict
+     
+        self.unsearched_urls = [url]
+        self.searched_urls = []
+        self.number_searched = 0 #counts number of pages searched.
+        
+        while self.unsearched_urls and self.number_searched < self.max_number_searched:
+            self.search_page()
+            print(list(self.keyword_dict.keys()))
+        
+    def compare_url_to_init(self,url):
+        init_url = self.init_url.split('/')
+        url_start = url.split('/')[:len(init_url)]
+        if '/'.join(url_start) == '/'.join(init_url):
+            return True
+        else:
+            return False
+        
+    def search_page(self):
+        url = self.unsearched_urls.pop(0)
+        print(url)
+        self.searched_urls.append(url)
+        self.number_searched += 1
+        
+        try:
+            page = requests.get(url)
+            html = BeautifulSoup(page.content,'html.parser')
+        except:
+            print('Unable to connect to {}'.format(url))
+            return
+        
+        self.scan_for_hits(html,url)
+        links = self.scan_for_links(html,url)
+        self.add_links_to_unsearched(links)
+
+    def scan_for_hits(self,html,url):
+        regex_dict = self.regex_dict
+        for keyword in list(regex_dict.keys()):
+            rgx = regex_dict[keyword]
+            match = html.find(string=rgx)
+            if match:
+                new_list = self.keyword_dict.get(keyword,[]) + [url]
+                self.keyword_dict[keyword] = new_list
+            
+    def scan_for_links(self,html,url):
+        link_tags = html.find_all('a')
+        links = []
+        for tag in link_tags:
+            href = tag.attrs.get('href',None)
+            if href is None:
+                continue
+            else:
+                if html.find('base'):  #search for base url for relative paths
+                    base = html.find('base').attrs.get('href','')
+                    if len(base)==0:
+                        base = url
+                else:
+                    base = url
+                try:
+                    link = requests.compat.urljoin(base,href) #join relative path. If href full path, then outputs href
+                except:
+                    link = href
+                link = self.cleanup_url(link)
+                if self.compare_url_to_init(link):
+                    links.append(link)
+        return links
+
+    def add_links_to_unsearched(self,links):
+        unsearched_urls = self.unsearched_urls
+        searched_urls = self.searched_urls
+        for link in set(links):
+            link = link.strip('/')
+            if link not in unsearched_urls and link not in searched_urls:
+                unsearched_urls.append(link)
```

Comment:
`searched_urls` and `unsearched_urls` could be sets. That way searching whether they contain a url is faster(linear search vs hashing), and also you don't need to check whether a url is in unsearched_urls before just adding it to the set.

```suggestion
            if link not in searched_urls:
                unsearched_urls.add(link)
```